### DISCUSSION TITLE
What do you use for a Markdoc markup compiler (serializer)?

### DISCUSSION THREAD

**Discussion by alex-sherwin** [markdoc/markdoc#366] [General]
I have a need to take structured content (think rich text editor AST, etc) and compile (serialize) into Markdoc markup.

To the best I can tell, I don't see anything in this repo that helps with generating/compiling/serializing Markdoc, so I was curious what Stripe does internally to create Markdoc markup programmatically (or maybe you don't do that at all?)

The route I've taken is to implement an extension for remark / mdast / mdast-util-to-markdown which adds AST support for generic Markdoc block and inline AST nodes.

I've implemented a fairly simple (hopefully not too naive) serialization, which really only has to focus on the actual Markdoc tag strings themselves, delegating back to the remark framework to recursively render child content.

Any thoughts on the matter would be greatly appreciated!

Thanks,

**Comment by matv-stripe** [comment]
Hi @alex-sherwin! If I understand your question correctly, it is about generating Markdoc and how to serialize them. We are experimenting with some of this right now were we are generating Markdoc nodes programmatically and serializing them as text files.

For example:
```
const document = new Ast.Node('document', {}, [
new Ast.Node('heading', { level: 2 } ), ['Hello world'])
])


const content = Markdoc.format(document);

// serialize content as a text document
```

It should also be possible to serialize a Markdoc AST into a file, and then use https://github.com/markdoc/markdoc/blob/ce14a807a06b97d4b4d0575daa356addb06a1cd0/src/ast/index.ts#L20 `fromJSON` to deserialize.

Let us know if that answers your question!

**Reply by alex-sherwin** [reply]
🤦‍♂️

Hi @matv-stripe, thanks for that.. really not sure how I missed it (for the record, that format function does not appear to be documented in the public facing docs)

I just checked out the code a bit and it appears to be a home grown Markdown + Markdoc serializer using the markdoc-defined AST object graph (which, is, basically exactly what I was looking for)

I do have a question, though.. how thoroughly vetted is the Markdown AST node serialization versus a battle-tested compiler, like remark, for example?

Now that I've gone through the trouble of adding Markdoc tag serialization to remark, I might stick with that since remark has extensively tested Markdown serialization techniques

**Reply by matv-stripe** [reply]
@alex-sherwin `Markdoc.format` is a relatively new feature. We have tested it across our doc corpus and it has been working (cc @mfix-stripe to double check me on that).

**Reply by mfix-stripe** [reply]
That's right! There might be very small edge cases, but nothing that I know of has cropped up since shipping the production version (when we renamed the function from `__EXPERIMENTAL__format` to `format`)

**Reply by alex-sherwin** [reply]
So, as it turns out... I went pretty far down the road using a remark based compiler (with an extension to serialize markdoc block/inline tags),

However, I ended up with a blocker.. unfortunately, you can't customize the remark behavior with how it serializes sequences of inline content, and it always ends all sequences of inline content with a newline, which means It would render an inline Markdoc tag, for example, like this:

```markdown
some paragraph {% strike %}striked content
{% /strike %}
```
Now, I think this actually is valid Markdown (this sequence should be a single inline sequence, despite the newline), however, the **Markdoc** parser doesn't handle this properly.

This *might*, arguably, be a Markdoc bug/edge-case around detecting inline tags/content...

But I'm just going to switch to building the Markdoc AST nodes graph directly and using the Markdoc format function on that, which should alleviate my issue

**Reply by matv-stripe** [reply]
@alex-sherwin That is an interesting edge case. I think I agree with you that the sequence should be treated as a single sequence of inline content.

**Reply by alex-sherwin** [reply]
I will put together a repro and look at it a bit more closely to see what you think.

It may be worth opining over, if it truly should work based on the formal Markdoc grammar, it may be worth fixing so that the parser would interface smoother with other ecosystem solutions (remark/unist has a large established suite of transforms and having markdoc in here as I'm trying would be quite nice)

**Reply by alex-sherwin** [reply]
The grammar spec https://markdoc.dev/spec#sec-Inline-form states

> When the [opening tag](https://markdoc.dev/spec#opening-tag) and [closing tag](https://markdoc.dev/spec#closing-tag) appear on the same line within a paragrah

So the current behavior is in line with the grammar description

So, I guess it's a philosophical question, since in Markdown the given example is a sequence of inline content (or if it were ended with a backslash or double spaces), whether the Markdoc parser should identify complete inline content sequences before scanning for inline open/close tags or just leave well enough alone.

My personal opinion is that it would be ideal if it would play nice with Markdown's notion of inline content sequences that are allowed to span multiple lines, thus making the Markdoc grammar play nice with other ecosystem tools that may do things like in my example (outside of the end-users control, as it's not customizable behavior)
